crates/sui/src/ethereum_client_commands.rs

```diff
@@ -232,3 +258,359 @@ pub(crate) async fn init_ethereum_state(
+
+    let mut pt_builder = ProgrammableTransactionBuilder::new();
+    client
+        .transaction_builder()
+        .single_move_call(
+            &mut pt_builder,
+            SUI_SYSTEM_PACKAGE_ID,
+            ETH_DWALLET_MODULE_NAME.as_str(),
+            APPROVE_MESSAGE_FUNC_NAME.as_str(),
+            Vec::new(),
+            Vec::from([
+                SuiJsonValue::from_object_id(eth_dwallet_cap_id),
+                SuiJsonValue::from_object_id(dwallet_id),
+                proof_sui_json,
+            ]),
+        )
+        .await?;
+
+    let tx_data = client
+        .transaction_builder()
+        .finish_programmable_transaction(sender, pt_builder, gas, gas_budget)
+        .await?;
+
+    let session_response = serialize_or_execute!(
+        tx_data,
+        serialize_unsigned_transaction,
+        serialize_signed_transaction,
+        context,
+        Call
+    );
+    Ok(session_response)
+}
+
+fn serialize_object<T>(object: &T) -> Result<SuiJsonValue, anyhow::Error>
+where
+    T: ?Sized + Serialize,
+{
+    let object_bytes = bcs::to_bytes(&object)?;
+    let object_json = object_bytes
+        .iter()
+        .map(|v| Value::Number(Number::from(*v)))
+        .collect();
+    Ok(SuiJsonValue::new(Value::Array(object_json))?)
+}
+
+async fn get_object_bcs_by_id(
+    context: &mut WalletContext,
+    object_id: ObjectID,
+) -> Result<SuiRawData, anyhow::Error> {
+    let object_resp = context
+        .get_client()
+        .await?
+        .read_api()
+        .get_object_with_options(
+            object_id,
+            SuiObjectDataOptions::default().with_bcs().with_owner(),
+        )
+        .await?;
+
+    match object_resp.data {
+        Some(data) => Ok(data.bcs.ok_or_else(|| anyhow!("missing object data"))?),
+        None => Err(anyhow!("Could not find object with ID: {:?}", object_id)),
+    }
+}
+
+async fn get_shared_object_input_by_id(
+    context: &mut WalletContext,
+    object_id: ObjectID,
+) -> Result<SharedInputObject, anyhow::Error> {
+    let object_resp = context
+        .get_client()
+        .await?
+        .read_api()
+        .get_object_with_options(object_id, SuiObjectDataOptions::default().with_owner())
+        .await?;
+
+    match object_resp.data {
+        Some(_) => {
+            let owner = object_resp
+                .owner()
+                .ok_or_else(|| anyhow!("missing object owner"))?;
+            let initial_shared_version = match owner {
+                Owner::Shared {
+                    initial_shared_version,
+                } => initial_shared_version,
+                _ => return Err(anyhow!("Object is not shared")),
+            };
+            Ok(SharedInputObject {
+                id: object_id,
+                initial_shared_version,
+                mutable: true,
+            })
+        }
+        None => Err(anyhow!("Could not find object with ID: {:?}", object_id)),
+    }
+}
```

Comment:
these seems like a useful functions outside of the eth-lc code, perhaps move it somewhere for other lightclients to use

Reply:
Done